### ISSUE TITLE
Discussion: level of backward compatibility with the legacy mxgraph codebase? 

### ISSUE THREAD

**Issue by tbouffard** [maxGraph/maxGraph#8]
My idea is to centralize decisions on this area and make any architectural changes documented and explicit. Feel free to comment.

For new mxGraph users, compatibility is not an issue, but for existing ones, introducing too much changes in one step could be challenging.

Starting point: https://github.com/jsGraph/mxgraph/issues/1#issuecomment-807982121

> I don't intend to maintain backwards compatibility for all of the modules as I think there needs to be some architectural changes to keep the codebase maintainable going into the future, but the structure+functionality of mxGraph overall seems fairly well thought out as-is in my opinion, even if it uses an older JS code style. I think I will probably keep the class names and function parameters etc the same for the most part.
> 
> Notably, I have already removed support for VML/Internet Explorer and Edge before it shifted to the Chromium engine, as I think that significantly eases maintenance.

**Comment by tbouffard** [comment]
I am ok for the `VML/Internet Explorer and Edge before it shifted to the Chromium engine` removing. This was historically kept for paying customers of the former maintainers of mxGraph. There is less and less demand for these old browsers, and as a community, this will slow us down.

As an existing user of mxGraph, that also uses it with the [mxGraph Typescript support](https://github.com/typed-mxgraph/typed-mxgraph), I need more details about @mcyph `I don't intend to maintain backwards compatibility for all of the modules` ideas.
I am opened to any changes but I would be happy to know which parts of mxGraph will be impacted and what ECMAScript version will be targeted (for now, comments in https://github.com/mcyph/mxgraph/blob/af8338c077478390b9fe588c075dfab93acce5e0/src/mxgraph/util/mxAbstractCanvas2D.js suggest that mxGraph may target ES9)

**Comment by mcyph** [comment]
There's a lot to cover here and I'm sure I'll miss some things. I'll need to add to this but I've tried to summarize what I've changed and what I'll be proposing to change from here.

Ideally, I think TypeScript support might be best integrated into the codebase too(?) I might need to read some more/ask for guidance on how to do this as I haven't used TypeScript before, but I have created type definitions using e.g. Cython before and think it might be similar.

---

Although I've made quite extensive changes to the codebase so far, I think so far there isn't a huge amount which would need to be changed for existing users of mxGraph aside from things like the way modules currently need to be explicitly `import`ed. A bundle with everything in one could probably be created to ease transition. I currently am using ES9 syntax like `for (const i of myList) {}`, although only in some areas and am using Babel to transpile, which should mean a vast majority of current browsers (Firefox/Chrome/Safari/Edge) would be supported.

Most of the changes I've made are things like:

* removing VML/IE6-11 support as mentioned (though IE9-11 might still work with some issues)
* conversion to ES6 classes from prototypes
* using ESLint to automatically arrange the code
* moving from `var` to `let`/`const`
* using `===`/`!==` instead of `==`/`!=` in cases where I thought it wouldn't cause side effects. I notice that `mxUtils.getValue` is also used extensively for configuration, but I was hesitant to use the triple operators in this case, and some of the other `mxUtils` functions also almost certainly rely on the behaviour of `==` and `!=` for comparisons.
* removing of `mxUtils.bind()` and replacing with `() => {}`-style prebound functions
* separating things so there's only a single class per file
* updating example code to use classes, local rather than global configuration and removing boilerplate (this actually has been a fairly large percentage of the work I've been doing)

There sometimes will be bugs I've introduced as a result of these changes, but I think for the most part these are back-compatible.

The way `mxClient` fits into everything and how e.g. the localization resources should be loaded needs to be clarified/fixed. Currently this functionality isn't working. Clarification about how the `mxCodecs` system works and how it integrates with tree-shaking also needs to be done at some stage, as the codecs aren't registered until the module is `import`ed. 

---

In terms of things I'm considering but haven't done yet, and may cause issues for backwards compatibility:

* I'm considering whether there's a way to prioritize [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance):
  * Need to figure out what to do with variables created in `mxGraph.createHandlers()`: `tooltipHandler`, `selectionCellsHandler`, `connectionHandler`, `graphHandler`, `panningHandler`, `popupMenuHandler`. (I personally think allowing instances as arguments to the `constructor` could be a better design than requiring subclasses to override e.g. `createConnectionHandler` for instance).
  * I'm thinking about whether there's a means of specifying configurations for different class instances, and not making configuration global for all future unrelated instances, with the goal of increasing flexibility and reducing side effects/bugs. It could mean another `config` variable to many classes, but I think the benefits could outweigh the downsides if there aren't better alternatives.

* If replacing the configuration system, we'll need to consider whether setting global options via the prototype as before will be supported via a backwards compatibility mechanism.

* Another thing to think about is, should changes to config objects take effect on instances that config has previously been associated with via "subscriptions"?

This might be further in the future, but I'd like to consider e.g. allowing for integrating charts, canvases and and other HTML content inside the SVG as nodes, possibly via foreignObject tags or the native SVG renderer in some chart systems like eCharts and hoping clarifying how everything fits together could eventually allow for this.

**Comment by mcyph** [comment]
Some potential ways of providing the config I've been considering are:

* With standard objects (may be simplest and most obvious)
* Using [Javascript `get`/`set` class accessors](https://www.w3schools.com/js/js_object_accessors.asp) like `o.targetPoint = x` or `o.targetPoint`
* Using a [JavaScript `Proxy`](https://www.javascripttutorial.net/es6/javascript-proxy/) 
* Explicit getters/setters like `o.setTargetPoint(x)`/`o.getTargetPoint()`

I'm currently leaning towards using JavaScript `Proxy`s as they could try explicit configuration via `config` argument first, `prototype` compatibility second, then default values last. They could be flexible with potential validation of values, and notification when values are changed to subscribing class instances. I'm also thinking it might need less documentation than if every variable had a separate `setX` or `getX` method. The downside I can see is it's a less used (though apparently widely supported) JavaScript feature that would be more complex at a lower level.

If doing this, I'll need to go through each variable in each class and document which are instance variables that shouldn't be in the configuration (like the `graph` instance), those which are configuration values and need to be moved, and the ambiguous ones that come somewhere in between of the two.

**Comment by junsikshim** [comment]
I've been trying to modernize mxgraph on my own, and I also had similar thoughts to mcyph's comments.

> Explicit getters/setters like o.setTargetPoint(x)/o.getTargetPoint()

I tried to hide the inner states inside a closure, and expose only the getters and setters. But while doing that, I realized there are so many properties created on the fly on `this`. Some of them are used only inside that component, so it would be better to not expose them to outside.

Moreover, in `control.html` example, an `mxImageShape` is created and assigned to a `mxCellState` as a new property named `deleteControl` which is not an override. It might be fine to use custom properties like that, but if people's implementations are based on some inner states(this.xxx is always available to the outside), hiding them will cause trouble.

Also, in `mxGraphHandler`, there is already a method called `getCells()` along with `this.cells = ...` assignments. So, I had to create something like `_getCells()` for internal use which wasn't very pretty.


> I'm considering whether there's a way to prioritize composition over inheritance:

I tried to throw out all the 'extends' stuff out, and use composition instead. So, extending `mxEventSource` became something like,
```js
const { fireEvent, addListener, removeListener } = mxEventSource();
```

It was working alright, but after finding out all the unexpected overrides, my efforts ran into a brick wall, especially with all the `mxShape` subcomponents. I was relying heavily on closures, but it was hard to mix and match with overriden methods.

Hope to discuss more and find a better way!

**Comment by mcyph** [comment]
Yeah sounds great! I think certainly over the longer term keeping the inner states hidden could be a good goal. 

My reasoning for using ES6 classes was not primarily because of inheritance but ability to keep the `constructor` and inheritence chain (the `extends ...`) and relevant variables at the top of each file so I could see the most important info when I opened them. ES6 classes have in my previous projects behaved not exactly the same, but very similar to prototypes, and I thought that using them could allow for incremental changes. I also think while they aren't well-supported at the moment, [private class fields](https://www.sitepoint.com/javascript-private-class-fields/) are likely to soon be widely available by the time the refactor is done, and could convert internal variables over time. 

As you say many of the examples set properties in the `prototype` of each class, and also frequently reference variables like `graph.container`. I've tried to convert many of the examples to use inheritance so they don't pollute the global namespaces, but my feeling is it may be best to keep things working the old way for at least a while until users of the library have had a chance to migrate over. (I'm mainly using inheritance there because of composition not currently being possible with how mxGraph et al are instantiated btw). It's tempting to go further in cleaning things up, but I think that way we would have more time to get feedback from people who will use the modified mxGraph with legacy codebases.
 
My feeling is so long as there's not too many levels of abstraction (preferably there would be as few as possible in my opinion) inheritance for shapes isn't a huge problem. But if there's lots of levels it can be hard to figure out what functionality/variables are where, and it might be easier to just duplicate methods. Like for instance mxCloud inherits from mxActor and mxShape, but perhaps it would be easier to only derive from a single class? I think perhaps it might be best to focus on other things first, but it may pay to use a diagramming tool (https://www.diagrams.net/ would be perfect 😄) to figure out how things interrelate and how to streamline things.

Maybe it's somewhat a radical idea, but I wonder whether more-or-less mimicking how css works with separation of style, inheritance and class names could be a good idea. Like for instance someone could do something like the below with JS accessors and/or `Proxy`s (note it's not taking into account the arguments/feasibility etc, it's just semi-pseudocode):

    const styles = {myClassName: {border: "1px solid blue"}};
    const myGraph = new mxGraph(div, styles);
    const myEllipse1 = new mxEllipse(className="myClassName");
    const myEllipse2 = new mxEllipse(className="myClassName");
    myGraph.appendChild(myEllipse1);
    myGraph.appendChild(myEllipse2);

    // Some time later, setting the below will instantly only affect myEllipse1/myEllipse2 respectively:
    myEllipse1.style.borderWidth = "5px";
    myEllipse2.style.borderColor = "red"; 

I think it might be viable to mimic the way other elements work, so as to minimize the amount new users need to learn..but that could also affect backwards compatibility as the `style` variable is taken in many classes etc..it's still early days and need to consider a fair amount of things first 🤔

**Comment by mcyph** [comment]
@junsikshim I want to look through in some more detail but I notice you've gone much further than I have in separating the constants and reorganizing mxUtils. I think it would be nice to integrate this as there's not much reason I can see for having a big utility class, especially when it makes tree-shaking more tricky due to so many files importing it.

**Comment by junsikshim** [comment]
Right. When I used mxGraph in my projects before, the bloated bundle size a big problem, so enabling tree-shaking has been a primary goal of mine.

Some other design choices I was trying to implement include...

I removed all `null`s from the code base in order to use the default function parameter functionality.

```js
mxGraphModel.prototype.getChildCells = function (parent, vertices, edges) {
  vertices = vertices != null ? vertices : false;
  edges = edges != null ? edges : false;
```
becomes,
```js
mxGraphModel.prototype.getChildCells = function (parent, vertices = false, edges = false) {
```

Since all the `null` checks are using `!=` and `==`, so using `undefined` instead of `null` is not a problem, but not using strict inequality feels kind of dirty, so I created `isSet() / isUnset()` helper function to check for nullish values.

```js
export const isSet = (v) => v !== undefined && v !== null;
export const isUnset = (v) => !isSet(v);
```

With that, `if (a != null)` becomes `if (isSet(a))`.
Checking for `null` as well as `undefined` is for backward compatibility.

Also, there are so many occurrences written like below,
```js
if (this.state.text != null &&
      this.state.text.node != null &&
      this.state.text.node.nextSibling != null) ...
```

It would be a lot better to take advantage of optional chaining and write it like,
```js
if (isSet(this.state.text?.node?.nextSibling))
```

I'll stop here before this comment gets too long :)

**Comment by junsikshim** [comment]
@mcyph

> Like for instance mxCloud inherits from mxActor and mxShape, but perhaps it would be easier to only derive from a single class?

I definitely think it should have only one super class. Extending from multiple classes doesn't make sense anyway, and if we treat them as mixins which we can compose, we might lose backward compatibility. In order to get more insights, I agree that we need some kind of diagram to see how things interrelate. But, I'm not sure where to start, if someone can begin to draw, I'd be willing to add on to it.

**Comment by mcyph** [comment]
@junsikshim yeah there's a lot to consider and I'd like to get my repo status up to the stage where I can potentially submit a PR+merge with jsGraph's repo so more people can contribute and I'm not holding anyone up. Main things which I need to do I can think of:

* Remove the React dependency, replacing with a basic node-only examples interface. Auto-generated examples output as .html could be possible once they're stable+is a good goal though it'd be harder to debug if needing to keep rebuilding when developing.
* Make mxUtils use your refactored version. (The rest of the changes you've made would be quite broad so debating whether to try to go through and merge all at once now or do a big PR later, but it'd be good if some of the typing problems could be resolved/reduced. If all else fails too I think putting our two codebases side by side and comparing/merging is an option).
* Figure out how to autogenerate the documentation.

I should note while I'm thinking about it that I've removed the .net and java versions of mxGraph too (in case users will be affected by this), but from where I'm standing it should be feasible to get something of mergeable quality soon, even if only to a dev branch. I think it'd be easiest to do architectural changes like changing the superclasses iteratively after merging, autogenerated class diagrams might be enough to show class relationships depending how we go as some IDEs have UML generation systems.

**Comment by junsikshim** [comment]
@mcyph Sounds great! I think we should start to build based on your work.

My version of refactoring was an experiment to apply some composition feels to it, but after weeks of trying I realized it is impossible with the current architecture of mxgraph. So, just disregard my work, and I'll contribute where I can :)

**Comment by tbouffard** [comment]
First, thanks @mcyph and @junsikshim for having taken time to share so much information and ideas. It is really great to see enthoustic people willing to improve the old venerable mxGraph library.
I will check your personal repository on next week to better understand the technical changes you propose.
Here are my 2 cents for the new mxgraph :smile:


> As you say many of the examples set properties in the prototype of each class, and also frequently reference variables like graph.container. I've tried to convert many of the examples to use inheritance so they don't pollute the global namespaces, but my feeling is it may be best to keep things working the old way for at least a while until users of the library have had a chance to migrate over. (I'm mainly using inheritance there because of composition not currently being possible with how mxGraph et al are instantiated btw). It's tempting to go further in cleaning things up, but I think that way we would have more time to get feedback from people who will use the modified mxGraph with legacy codebase

This is exactly what I had in mind: let's have a large goal for the mxGraph internals "be ready to change anything internally" but do it at a regular pace to avoid
- spending months developing a new version without it being used
- breaking the whole usage of the lib for existing users when it is ready
 
Let's define the baby steps for the internal changes that will also bring new capabilities for new features like some already expressed in comments on this issue.

As a mxGraph user, what I see in the first part of https://github.com/jsGraph/mxgraph/issues/8#issuecomment-808653862 (the list under 'Most of the changes I've made are things like') seems not a big deal and the changes I expect to have to manage are not too big as well.
IMHO, it is ok to have to change the way the class are imported and the lib is instantiated (I will be so happy if we can remove all the globals and the way mxClient is spread all around), moving to ES6 classes neigher (I largely extend the mxGraph code in my current project, so I don't see it as an issue).

In the first step, we should limit the public API changes (for instance, avoid rename, yes prefixing everything by mx is crappy, but we can live with it at the beginning) except the removal already discussed here.
Anything else, like composition over inheritance, configuration changes, internal private class fields, .... should be done later

About the class diagrams you want to create, I have found this old resources (I didn't check in details, so this may be useless): https://github.com/process-analytics/mxgraph2/tree/mxgraph-4_1_1/etc/diagrams

**mobile support**
The lib previously worked on mobile, so we should keep the support as well. The mobile browsers list to support should be reviewed I guess.


**typescript support** 
As a mxGraph TypeScript user, having mxGraph written in TypeScript would be nice, but I am not sure it should be done as a first step. I don't have any idea of the effort needed to convert the new code into TS.
But at least we could provide the TypeScript definitions.
I guess we could benefit from the https://github.com/typed-mxgraph/typed-mxgraph experience (I am a contributor of this project) and insights https://github.com/typed-mxgraph/typed-mxgraph/issues/12.
The challenge we faced in typed-mxgraph with the legacy mxgraph codebase is that 'many properties are created on the fly on this' as @junsikshim noticed. Having the correct properties on types is a crafting work and we have missed a lot - most feedback we got from user is in this area.
So without internal changes in the mxGraph codebase, generating the types could be challenging.



**documentation API generation**
According to an archive of the old mxGraph development repository (https://github.com/process-analytics/mxgraph2/blob/mxgraph-4_1_1/build.sh#L104), the API documentation was previously generated with https://www.naturaldocs.org/
The https://jgraph.github.io/mxgraph/docs/js-api/files/index-txt.html page is generated from  https://github.com/jsGraph/mxgraph/tree/5dc3a07195313fd134bf8b058f4ae1d660202e07/javascript/src/js/index.txt
This tool uses a uncommon format comparing to JSdoc which is supported by most IDE, so we may switch to another API documentation format.

In https://github.com/typed-mxgraph/typed-mxgraph, we have started migrating the API documentation to TSDoc, which can produce html documentation with https://typedoc.org for instance.
See https://tbouffard.github.io/typed-mxgraph/api/index.html (no typedoc configuration, this could be improved if needed). 


**java and dotnet support**
IMHO, they should be removed from this repository and the code should be moved to dedicated repositories if someone is ready to maintain it. The java version hasn't been updated for a while. I personally only use it indirectly with tools like https://github.com/process-analytics/mxgraph-svg2shape or https://github.com/jgraph/svg2xml

**Comment by junsikshim** [comment]
Great summary!
I agree that it'd better to focus on reformatting the current code and keep breaking-changes minimal.

As with mobile, documentation, java & dotnet support, I totally agree with @tbouffard.

While it'd be great to have TS support, I think we should only provide the definitions. The more I look into the code, I feel like it would take months(if not impossible) to just get compiled with TS on :(

Here are some other details I'm thinking about.

- Support only svg. Then we can remove all vml/xml/html supports including (mxVmlCanvas2D, mxXmlCanvas2D, many methods and conditions and all the 'dialects'), but I'm not 100% sure if those are still needed in some cases.
- Remove mxXmlRequest.
- Remove mxResources. Do people use it?
- Remove IE CSS filters.
- Remove quirks mode support.

I guess we got to set the browser support policy first, then other decisions will follow naturally. I think IE11+ would be ok, but wonder what people are expecting.

**Comment by mcyph** [comment]
I've actually been making some pretty big changes in last few days which could have an impact on where to go from here, notably I've already converted a large % of the `view`, `shape` and some of the `util` modules to TypeScript.  I was initially hesitant to using TypeScript but as a result of doing it and IDE hints I've already managed to spot quite a few inconsistencies in the documentation and code, and have been able to understand the way everything interrelates by knowing whether e.g. something returns a `mxPoint` or `mxRectangle`. It also has the advantage of enabling making sure classes such as `mxCellRenderer` don't set properties on other class instances like `mxCell` that aren't documented.

I initially started experimenting with `get`/`set` class accessor methods in `mxGraphView` in the hope of adding some validation of the variables used by each class, but I'm no longer in favor of this idea, because it adds a lot of boilerplate and could break compatibility with some older code. I actually think that rewriting in TypeScript could take less time and result in higher quality overall, I think I'm only days away from getting to that stage. I don't think we need to rewrite everything in TypeScript though, it's mainly core things like the datatypes and `mxGraph`. Some SVG things have been tricky to convert, and converting `mxGraph` to TS has been challenging because of how large that file is (and in fact I'm not sure whether/whether not it's better to split that file up at this stage) but hoping can make it compile soon.

* I'd already basically started removing IE CSS filters/vml/quirks support and have essentially been removing all non-SVG code. I think we should consider more tightly integrating SVG, as we may be able to support styles/filters and other things that previously weren't possible when supporting VML at the same time.
* I've started removing accessing style values via mxConstants variables, instead favoring directly referencing the properties. My reasoning is that they're always the same and the names usually obvious so there's no need to refer to the keys externally. TypeScript can also enforce property setting/getting. For example, in `mxShape`:
    
  ```javascript
  this.fill = mxUtils.getValue(this.style, mxConstants.STYLE_FILLCOLOR, this.fill);
  this.gradient = mxUtils.getValue(this.style, mxConstants.STYLE_GRADIENTCOLOR, this.gradient);
  this.gradientDirection = mxUtils.getValue(this.style, mxConstants.STYLE_GRADIENT_DIRECTION, this.gradientDirection);
  this.opacity = mxUtils.getValue(this.style, mxConstants.STYLE_OPACITY, this.opacity);
  ```
  becomes
  ```javascript
  this.fill = ifNotNullElse(this.style.fillColor, this.fill)
  this.gradient = ifNotNullElse(this.style.gradientColor, this.gradient);
  this.gradientDirection = ifNotNullElse(this.style.gradientDirection, this.gradientDirection);
  this.opacity = ifNotNullElse(this.style.opacity, this.opacity);
  ```
* I'm on the fence about whether to remove the `mx` prefix from everything. On the one hand I'm not too fussed by the mx prefix, but on the other it'd be nice to remove it as some IDEs like PyCharm/WebStorm complain with a red underline that "constructors should start with an upper case letter" and compatibility could be provided by import aliases e.g. `import { Graph as mxGraph } from 'mxGraph/Graph'` or `import * from 'mxGraph/compat'`
* I've tested my previous changes (while the codebase was compiling) briefly on mobile and things should be OK. I don't think I've changed anything much that's likely to affect that, but I agree we should keep testing to make sure of this.
* It might be a good idea to look at your TypeScript project @tbouffard and see whether we can use your definitions to check I've annotated everything correctly, and also maybe copy the docstrings across. I quite like the look of something like https://typedoc.org/ so it'd be good to both correct inconsistent docs+update it at the same time. I don't think it'll necessarily be essential to do this to get things to release (or at least "dev" or "testing") quality, but it'd certainly be a good idea to do it once typescript conversion is further along.

I don't think this codebase is likely to be that unmaintainable going into the future, especially now most of it's been migrated away from prototypes+updated to newer syntax. I used to write code in the same sort of style/language features largely because of older JS+IE's limitations too, and it was very tricky to build on things, but JavaScript/ECMAScript and browsers have come a long way in the last few years. I think it would be feasible to keep this lib functional even with small updates, assuming standards like SVG don't have support removed/have backwards compatibility broken for some reason.

**Comment by mcyph** [comment]
@tbouffard the typed mxGraph project you contributed to has a lot of the type information already, and I think many of the remaining files are likely to be less dynamic than the core/types I've started to manually convert. I think it might be a good idea for me to automatically extract the docs and types from this project and add type info automatically when feasible.

**Comment by mcyph** [comment]
A common thing I've noticed throughout many files is code like the below in `mxGraphModel`:

```javascript
  filterCells(cells: mxCell[] | null,
              filter: Function): mxCell[] | null {

    let result = null;
    if (cells != null) {
      result = [];

      for (let i = 0; i < cells.length; i += 1) {
        if (filter(cells[i])) {
          result.push(cells[i]);
        }
      }
    }
    return result;
  }
```

I won't make it a goal for the first release, but I think a lot of these methods actually should be in `mxCell` or a `mxCellCollection` or something rather than having lots of "managing" classes. 

I also think it could be a good idea to not do as many checks like `if (cells != null) {` throughout the codebase. I'm hesitant to change this kind of code for now because it could break things for existing users, but ironically this could cause unintended bugs and my preference in future releases for many/most of these cases would be to have these functions explicitly "fail fast" rather than do nothing if `null` is passed to them and have to look through many layers to find out why something doesn't happen.

**Comment by mcyph** [comment]
Hmmmm, I've just been researching svg-related things like d3.js but was concerned it might be too large for our purposes (~70kb minified/compressed) and found [this article](https://medium.com/@PepsRyuu/why-i-no-longer-use-d3-js-b8288f306c9a).

I wonder whether using https://preactjs.com/ could be worth considering for both the examples and things like the shapes for reducing the amount of code? I think so long as we didn't use JSX like at http://jsfiddle.net/developit/e281k4wz/ the code could still work without needing node.js/babel etc and could be added directly to the browser.

I can see this could be a controversial decision, but in order to keep the examples to a minimum and make sure a release could be made sooner I was basically planning to do something "react-like but not react" in the sense they mostly are separated into the HTML template+logic code and thinking it could be good to consider this if it essentially does the same thing. Potentially I think this could also reduce the bundle size overall if classes could be made less verbose.

**Comment by junsikshim** [comment]
@mcyph I guess that `filterCells()` method can actually be a generic `filter()` method and moved to `mxUtils` or something. If a method is not using any of the inner state(`this`), it may be wise to move it out and encourage reuse. While on that subject, there are multiple functions in `mxUtils` that are not needed anymore such as `indexOf` and `forEach`. I think we should remove those functions are change the code accordingly.

As for the null checks, I agree there are too many of them throughout the codebase and they hamper the code readability. There might be some underlying problem just not visible because of that `if` statement.

Also, as with `filterCells()`, I prefer an array to be never a `null` and be an empty array instead. That way, we can apply multiple operations on that array without having null checks in between. However, that would return `[]` instead of `null` in that method, and we might lose some backward compatibility, but I think it's worth it.

D3 is a great library and I love using it, but I think the overall structures are so different that it would be near impossible to incorporate into one. I wonder what has brought you to search for libraries as it'd take more time to mix them than just converting it to a modern format. And I think it's pretty hard to read react/preact code without jsx. One of the best things about mxgraph is that it doesn't have any dependencies, so unless there's a really good reason I think we should avoid adding one. :)

**Comment by mcyph** [comment]
Yeah I don't have any plans to put preact into the core lib for this release. I think if that is a direction which is worth heading in that it may be best to fork as "mxGraph-next" and simply maintain the older one as browsers change for older apps, as things like removing all the null checks and the redundant methods would break many things either way.

I'm still not sure what to do about the examples for this release, which currently use react. The alternative to react is probably something vue-like with the template and code in separate files I guess(?) There's certainly advantages to that approach, so if you think that that's a good way to go (or you have alternative ideas too!) I'd be very interested.

I'm currently mainly researching svg+the svg ecosystem to figure out what it's capable of and how it differs from vml. Currently I'm reading the book "Using SVG with CSS3 and HTML5", as I was thinking that going forwards I would need to understand the way svg is supported and the best ways to use it, especially if mxGraph was to be extended with additional svg-related functionality.

I think the main things React/Preact were designed to solve were things like state, composability, and that only elements which change are updated. There's actually a huge amount of code right the way through mxGraph to only update changed aspects to keep things performing well. mxGraph also has quite a bit of serialization to/from XML. This has been tricky to convert to TypeScript because of allowing classes to be instantiated without defining some instance variables and I'm wondering whether this serialization performed by using a virtual DOM could reduce repetition and work around this kind of problem. Shapes rendering directly to SVG could also reduce the level of indirection. I wouldn't be surprised if around 1/3 of mxGraph's code could be reduced centralizing these kinds of functions.

**Comment by junsikshim** [comment]
@mcyph 
While I use React all the time at work, the bundle size of react-dom has always been a problem for me. Preact might be better, but I still can't figure out in my head how to integrate react/preact with mxGraph. I think we need to gather more thoughts on this subject.

Speaking of frameworks, many users will be using one of them. So, it may be wise to provide React/Vue/Angular/etc. wrappers while keeping the core vanilla-js. We could make a monorepo and have folders like `packages/react`, `packages/vue`. We could also add Storybook for each of them to provide interactive examples instead of the current static vanilla-only ones.

**Comment by mcyph** [comment]
@junsikshim I can absolutely understand your concerns - I won't do it here because it will be quite long but I think it would be best if I write up an example of how the library would be used with the previous way alongside the react/preact way so I can explain what I mean later on in the newly opened ["Discussions" tab](https://github.com/jsGraph/mxgraph/discussions).

I think if we put them side by side we could better analyze how the architecture could be designed, and also consider whether there are ways of achieving the same or similar things by not using react/preact.

I hadn't considered using these before now, I just recalled reading somewhere they were part of HTML5 but custom elements (or at least "autonomous custom elements") are apparently [supported](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) by all current browsers even if customized built-in elements aren't supported by Safari, so I'm wondering whether using these may be an alternative way of allowing for "de/serialization by default" while not requiring any external framework. There's some more comprehensive examples at https://developers.google.com/web/fundamentals/web-components/customelements and they seem quite powerful and may even have advantages over using an external library.

**Comment by junsikshim** [comment]
@mcyph I've never used web components in projects before, but I think it's a clever idea especially for the shapes. I wonder what kind of real-world problems/limitations are there. Hope to see someone sharing the experience.

**Comment by mcyph** [comment]
I think I'll leave this for now but I've pushed my changes (including conversion of the main base parts of mxGraph to TypeScript) to https://github.com/mcyph/mxgraph if anyone would like to look at the current status. It might be a few weeks before I can make bigger changes but I think it's a start, and that it may be a good idea to merge back to an "unstable" branch of jsGraph if anyone else would like to make changes in the meantime. 

The conversion to TypeScript has introduced some bugs which need to be fixed but I'm hoping these can be solved asap, and think in the longer term a TypeScript version should be significantly easier to build on+maintain. Ideally many/most type assertions and `@ignore-ts` comments should be removed over time but that will need `if ... != null` checks removed and types changed throughout many files.

I need to do a writeup in more detail but I've uploaded a demo of what web components could look like, together with SVG at https://gist.github.com/mcyph/3e91d76ef624e73c13e93b7dc15f27bc.

I've also written a script to extract the docs+types from typed-mxgraph, the output is at https://gist.github.com/mcyph/d6a19487506eaf7db7e95ccd9e487a6b. I'm thinking it may be best to do a big auto-update of the docs, maybe with some kind of levenshtein/diff algorithm to spot potential errors. Ideally the rest of the .js files should be converted to TypeScript too, but that will take longer and I think it's best we stabilize the existing code first.

**Comment by mcyph** [comment]
@tbouffard I've auto-added most of the type annotations and JSDoc comments from the typed-mxgraph project. With a few exceptions (like mxUtils and mxConstants) I think it's worked for a majority of cases. I've put the type definition as a `// ...`-style comment underneath the docstring to allow easier checking of the converted .ts files, and hopefully also it can make things easier to convert for the .js files which haven't yet been converted to .ts. I'm actually pretty amazed how comprehensive the typed-mxgraph docs+types are!

**Comment by junsikshim** [comment]
@mcyph How about pushing to a branch here so we can all work on it? Possibly starting with custom webpack settings instead of using next. The folders might have to be rearranged though (examples and so on.)

**Comment by mcyph** [comment]
@junsikshim Will need to check whether I can be added as admin or have an admin create a branch here but yeah sure!

Cleaning up the examples and removing react/next is a substantial task and is probably one of, if not the largest thing that still needs to be done. I won't have time for the next few weeks to do this, but I think it's worth noting the package.json for the react/next examples is in `./src/package.json`, whereas the library itself is in `./package.json`, which means that users of the library wouldn't need to download react/next.

I started trying to remove react+next but needed to revert my changes so that I could work on refactoring the core part of mxGraph. I'm still not sure the best way of doing this - some of the most complex examples are in the `./docs/stashed` directory and I don't know whether it would be best to get the rest of the examples working stably first, or whether there's some way of removing react first and still being able to debug+easily build on them. I think actually the biggest task is refactoring the examples and removing react shouldn't take as much time. If you'd like to work on this in the meantime that'd be great, but either way my goal still remains to remove react+next when it's practical to do so.

**Comment by junsikshim** [comment]
@mcyph I think I can help to setup project and examples while you work on the core itself. I haven't looked at your refactored examples yet, but once the dev setup is complete, it'll be much easier to work with. Also, I think it's useful to have multiple versions(vanilla, react, etc.) of the same example probably with proper wrappers so people don't have to wrap them manually with `ref`s. Anyway, let's put it into a branch somehow and begin working on it :)

**Comment by mcyph** [comment]
@junsikshim Sounds great - I've added some explanations about how the examples are structured to https://github.com/mcyph/mxgraph/blob/master/src/pages/README.md, please let me know if there's anything you'd like me to clarify etc!

**Comment by mcyph** [comment]
Currently there's a bug I'm having trouble tracking down where if you move the node/vertices they don't update the position of connected edges. I'm not sure whether it's because of a changed default argument or a different change. [**EDIT:** Fixed the bug, it turned out to be to do with the `cell.invalidating` variable not being reset in `mxGraphView` and not to do with this, but leaving the below here as it could have consequences in other situations]

Supplying `null` won't go to a default anymore in most cases because of the functions using default arguments, e.g. previously code looked like this:

```javascript
function myFn(myArg) {
    myArg = myArg != null ? myArg : true;
}
```

Whereas now it looks like this with TypeScript/default arguments:

```javascript
function myFn(myArg: boolean=true) {
    ...
}
```

This could have further ramifications in some cases for existing codebases which expect `null` to use the default argument. I'm hoping there isn't too much code like this in the wild as default parameters are a lot cleaner but it's hard to say.

One thing that's very tempting going forwards is to separate `mxCell` into different `mxNode` (or `mxVertex`) and `mxEdge` classes, as I think it would make things more idiomatic and allow putting edge- or node-specific functionality into the types. That may also break code, but it's hard to say how much because I think a lot of the time external code doesn't normally directly use `mxCell`.

**Comment by junsikshim** [comment]
I agree that it'd better to separate mxCell into mxVertex and mxEdge(the handlers are already separated into mxVertexHandler and mxEdgeHandler). As @mcyph said, I also don't think it is a huge architectural change so it could be done in near future.

**Comment by mcyph** [comment]
🤔 yeah the more I look at the codebase the more I feel that could be one of the best ways of making the source more easily navigable for new contributors and being able to build on things. It would have some effect on compatibility but it would probably only need relatively minor changes for those already using the library. 

I wonder whether `mxCell` and `mxCellState` should be merged prior to doing this, and the functions of `mxGraphModel` largely merged into either `mxVertex` or `mxEdge` too. In fact, a large number of the functions which take `mxCell` or `mxCellState` as their first parameter could be merged into the types and just create forwarding functions for back-compat in the original classes.

 I'm trying to figure out whether there's code where `mxCell`/`mxCellState` would need to be separate, like in the undoable transactions code, but even there I'm not sure and in those cases it might be possible to work around by cloning all the properties each time.

**Comment by junsikshim** [comment]
I'm not sure where it was, but I think I read something like, `mxCell` is the actual data structure and `mxCellState` works like a cache. And when the change is not yet made(while dragging or something like that), only `mxCellState` is updated.

So, since their roles are divided into a model and a view, I think it's better to keep them separated even if there are lots of overlapping features. I may be wrong as I don't have a thorough understanding of the architecture though.

**Comment by mcyph** [comment]
Yes, I think you're right. The mxCell docs says it "represents the state of the groups, vertices and edges in a graph" and the mxCellState "represents the current state of a cell in a given `mxGraphView`", so it's sort of like the state of the cell states I guess 🤔  In that case maybe the cell states should be stored in the graph view?? 

I've created a throwaway branch+organized all the files based on what the first parameter to each function is. I'm not sure what will come of this but essentially these classes seem like semi-singleton utility classes, and think in many cases they would be easier to work with if they had a mxCell or mxCellState associated with them as the first or last parameter of their constructor. (If it was the last, we could keep back-compat with the old way, though it wouldn't be the optimum having to pass `null` as the first param in every class call for new code.) I'm trying to think how this fits into the [SOLID principles](https://en.wikipedia.org/wiki/SOLID) and what consequences this has for maintenance and understandability.

![image](https://user-images.githubusercontent.com/20507948/114386574-beb77c00-9bd4-11eb-985c-c4560de39059.png)

**Comment by tbouffard** [comment]
Closing
Discussions continue in #18, #19, #41 and #51
